File: samples/python/text_generation/compound_grammar_generation.py
#!/usr/bin/env python3
# Copyright (C) 2025 Intel Corporation
# SPDX-License-Identifier: Apache-2.0

import argparse
import json
from typing import Any

from openvino_genai import (
    GenerationConfig,
    LLMPipeline,
    StreamingStatus,
    Parser,
    IncrementalParser,
    TextParserStreamer,
    DecodedResults
)

from openvino_genai import (
    StructuredOutputConfig as SOC,
)
from pydantic import BaseModel, Field


def streamer(subword):
    print(subword, end="", flush=True)
    return StreamingStatus.RUNNING


class book_flight_ticket(BaseModel):
    """booking flights"""

    origin_airport_code: str = Field(description="The name of Departure airport code")
    destination_airport_code: str = Field(description="The name of Destination airport code")
    departure_date: str = Field(description="The date of outbound flight")
    return_date: str = Field(description="The date of return flight")


class book_hotel(BaseModel):
    """booking hotel"""

    destination: str = Field(description="The name of the city")
    check_in_date: str = Field(description="The date of check in")
    checkout_date: str = Field(description="The date of check out")


def _recursive_purge_dict_key(d: dict[str, Any], k: str) -> None:
    """Remove a key from a dictionary recursively"""
    if isinstance(d, dict):
        for key in list(d.keys()):
            if key == k and "type" in d.keys():
                del d[key]
            else:
                _recursive_purge_dict_key(d[key], k)


def tool_to_dict(tool: BaseModel, with_description: bool = True) -> dict[str, Any]:
    schema = tool.model_json_schema()
    _recursive_purge_dict_key(schema, "title")
    if not with_description:
        _recursive_purge_dict_key(schema, "description")
    return {
        "type": "object",
        "properties": {
            "name": {"type": "string", "enum": [tool.__name__]},
            "arguments": schema,
        },
        "required": ["name", "arguments"],
    }


def tools_to_array_schema(*tools: BaseModel) -> str:
    return json.dumps(
        {
            "type": "array",
            "items": {"anyOf": [tool_to_dict(tool, with_description=False) for tool in tools]},
        }
    )


class IncrementalToolCallParser(IncrementalParser):
    """Incremental parser to extract tool calls from the model output.

    Custom parser should be inherited from IncrementalParser and implement 'parse' and 'reset' methods.
    """
    def parse(self, msg: dict, delta_text: str, delta_tokens: list) -> str:
        if 'content' not in msg:
            msg['content'] = ''
        msg['delta_text'] = delta_text

        # Join the previous content with the new delta_text for searching.
        # Do not modify msg['content'] yet, as it will be updated automatically by the streamer.
        content = msg['content'] + delta_text
        
        start_tag = "functools"
        start_index = content.find(start_tag)
        if start_index == -1:
            return delta_text

        msg['generates_tool_call'] = True
        if not content.endswith("}]"):
            return delta_text
        json_part = content[start_index + len(start_tag):]
        try:
            tool_calls = json.loads(json_part)
            msg['tool_calls'] = tool_calls
            msg['generates_tool_call'] = False
            return delta_text
        except json.JSONDecodeError as e:
            return delta_text
    
    def reset(self):
        # This method should be implemented in inherited classes.
        # e.g. self.text_cache = ""
        # But since in this implementation no states were used so do nothing.
        print('Parser state has been reset.')

class CurrentTextParserStreamer(TextParserStreamer):
    """
    A TextParserStreamer that receives parsed dictionary every time new text is generated.

    In order to get get parsed dictionary from the model output, a custom implementation of TextParserStreamer 
    with defined 'write' methods is needed.
    """
    def write(self, msg: dict):
        # If the tool call is not yet complete, continue streaming
        print(msg['delta_text'], end="", flush=True)
        return StreamingStatus.RUNNING


def print_tool_call(answer: DecodedResults):
    for tool_call in answer.parsed[0]['tool_calls']:
        print(f"{tool_call['name']}({', '.join(f'{key}=\"{value}\"' for key, value in tool_call['arguments'].items())})")


# modified system message from:
# https://github.com/vllm-project/vllm/blob/main/examples/tool_chat_template_phi4_mini.jinja
sys_message = """You are a helpful AI assistant.
You can answer yes or no to questions, or you can chose to call one or more of the provided functions.

Use the following rule to decide when to call a function:
    * if the response can be generated from your internal knowledge, do so, but use only yes or no as the response
    * if you need external information that can be obtained by calling one or more of the provided functions, generate function calls

If you decide to call functions:
    * prefix function calls with functools marker (no closing marker required)
    * all function calls should be generated in a single JSON list formatted as functools[{"name": [function name], "arguments": [function arguments as JSON]}, ...]
    * follow the provided JSON schema. Do not hallucinate arguments or values. Do not blindly copy values from the provided samples
    * respect the argument type formatting. E.g., if the type is number and format is float, write value 7 as 7.0
    * make sure you pick the right functions that match the user intent
"""


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument(
        "model_dir",
        help="Path to the model directory. It should contain the OpenVINO model files.",
    )
    args = parser.parse_args()

    pipe = LLMPipeline(args.model_dir, "CPU")
    tokenizer = pipe.get_tokenizer()
    chat_history = [{"role": "system", "content": sys_message}]
    tools = [tool_to_dict(tool) for tool in [book_flight_ticket, book_hotel]]

    generation_config = GenerationConfig()
    generation_config.max_new_tokens = 300
    generation_config.do_sample = True

    user_text_1 = "Do dolphins have fingers?"
    print("User: ", user_text_1)
    chat_history.append({"role": "user", "content": user_text_1})
    model_input = tokenizer.apply_chat_template(chat_history, add_generation_prompt=True, tools=tools)
    # same as SOC.Union(SOC.ConstString("yes"), SOC.ConstString("no"))
    yes_or_no_grammar = SOC.ConstString("yes") | SOC.ConstString("no")
    generation_config.structured_output_config = SOC(structural_tags_config=yes_or_no_grammar)
    print("Assistant: ", end="")
    answer = pipe.generate(model_input, generation_config, streamer=streamer)
    chat_history.append({"role": "assistant", "content": answer})
    print()

    user_text_2 = (
        "book flight ticket from Beijing to Paris(using airport code) in 2025-12-04 to 2025-12-10 , "
        "then book hotel from 2025-12-04 to 2025-12-10 in Paris"
    )
    print("User: ", user_text_2)
    chat_history.append({"role": "user", "content": user_text_2})
    model_input = tokenizer.apply_chat_template(chat_history, add_generation_prompt=True, tools=tools)

    start_tool_call_tag = SOC.ConstString(r"functools")
    tools_json = SOC.JSONSchema(tools_to_array_schema(book_flight_ticket, book_hotel))
    tool_call_grammar = start_tool_call_tag + tools_json  # SOC.Concat(start_tool_call_tag, tools_json)
    generation_config.structured_output_config.structural_tags_config = tool_call_grammar

    print("Assistant: ", end="")
    custom_streamer = CurrentTextParserStreamer(pipe.get_tokenizer(), [IncrementalToolCallParser()])
    answer = pipe.generate([model_input], generation_config, streamer=custom_streamer)
    
    print("\n\nThe following tool calls were generated:")
    print_tool_call(answer)

    print()


if __name__ == "__main__":
    main()

File: samples/js/text_generation/compound_grammar_generation.js
// Copyright (C) 2025-2026 Intel Corporation
// SPDX-License-Identifier: Apache-2.0

import { z } from 'zod';
import {
    ChatHistory,
    GenerationFinishReason,
    LLMPipeline,
    StructuredOutputConfig as SOC,
    StreamingStatus,
} from 'openvino-genai-node';
import { serialize_json, toJSONSchema } from './helper.js';

function streamer(subword) {
    process.stdout.write(subword);
    return StreamingStatus.RUNNING;
}

const bookFlightTicket = {
    name: "book_flight_ticket",
    schema: z.object({
        origin_airport_code: z.string().describe("The name of Departure airport code"),
        destination_airport_code: z.string().describe("The name of Destination airport code"),
        departure_date: z.string().describe("The date of outbound flight"),
        return_date: z.string().describe("The date of return flight"),
    }).describe("booking flights"),
};

const bookHotel = {
    name: "book_hotel",
    schema: z.object({
        destination: z.string().describe("The name of the city"),
        check_in_date: z.string().describe("The date of check in"),
        checkout_date: z.string().describe("The date of check out"),
    }).describe("booking hotel"),
};

// Helper functions
function toolToDict(tool, withDescription = true) {
    const deleteDescription = (ctx) => delete ctx.jsonSchema['description'];
    const jsonSchema = toJSONSchema(
        tool.schema,
        withDescription
            ? undefined
            : { override: deleteDescription }
    );

    return {
        type: "object",
        properties: {
            name: { type: "string", enum: [tool.name] },
            arguments: jsonSchema,
        },
        required: ["name", "arguments"],
    };
}

function toolsToArraySchema(...tools) {
    return serialize_json({
        type: "array",
        items: {
            anyOf: tools.map(tool => toolToDict(tool, false)),
        },
    });
}

class CustomToolCallParser {
    constructor() {
        this.content = "";
    }

    write(deltaText) {
        this.content += deltaText;

        const startTag = "functools";
        const startIndex = this.content.indexOf(startTag);
        if (startIndex === -1) {
            return StreamingStatus.RUNNING;
        }

        const jsonPart = this.content.slice(startIndex + startTag.length);
        try {
            JSON.parse(jsonPart);
            return StreamingStatus.TOOL_CALL_STOP;
        } catch {
            return StreamingStatus.RUNNING;
        }
    }

    parse(msg) {
        if (!msg.content) {
            msg.content = "";
        }
        const content = msg.content;

        const startTag = "functools";
        const startIndex = content.indexOf(startTag);
        if (startIndex === -1) {
            return;
        }

        const jsonPart = content.slice(startIndex + startTag.length);
        try {
            const toolCalls = JSON.parse(jsonPart);
            msg.tool_calls = toolCalls;
            return;
        } catch {
            return;
        }
    }
}

class ContentStreamer {
    constructor(tokenizer, parsers = []) {
        this.tokenizer = tokenizer;
        this.parsers = parsers;
    }

    write(subword) {
        process.stdout.write(subword);
        for (const parser of this.parsers) {
            const status = parser.write(subword);
            if (status !== StreamingStatus.RUNNING) {
                return status;
            }
        }
        return StreamingStatus.RUNNING;
    }
}

function formatFinishReason(reason) {
    const effectiveReason = reason ?? GenerationFinishReason.STOP;
    return GenerationFinishReason[effectiveReason] ?? effectiveReason;
}

function printToolCall(answer) {
    for (const toolCall of answer.parsed[0].tool_calls) {
        const args = Object.keys(toolCall["arguments"])
            .map((key) => `${key}="${toolCall["arguments"][key]}"`);
        console.log(`${toolCall["name"]}(${args.join(", ")})`);
    }
}

// System message
let sysMessage = `You are a helpful AI assistant.
You can answer yes or no to questions, or you can choose to call one or more of the provided functions.

Use the following rule to decide when to call a function:
    * if the response can be generated from your internal knowledge, do so, but use only yes or no as the response
    * if you need external information that can be obtained by calling one or more of the provided functions, generate function calls

If you decide to call functions:
    * prefix function calls with functools marker (no closing marker required)
    * all function calls should be generated in a single JSON list formatted as functools[{"name": [function name], "arguments": [function arguments as JSON]}, ...]
    * follow the provided JSON schema. Do not hallucinate arguments or values. Do not blindly copy values from the provided samples
    * respect the argument type formatting. E.g., if the type is number and format is float, write value 7 as 7.0
    * make sure you pick the right functions that match the user intent
`;

async function main() {
    const modelDir = process.argv[2];
    if (!modelDir) {
        console.error("Please provide the path to the model directory as the first argument.");
        process.exit(1);
    }

    const pipe = await LLMPipeline(modelDir, "CPU");
    const tools = [bookFlightTicket, bookHotel].map((tool) => toolToDict(tool, true));
    const chatHistory = new ChatHistory();
    chatHistory.setTools(tools);
    chatHistory.push({ role: "system", content: sysMessage });

    const generationConfig = {
        max_new_tokens: 300,
        do_sample: false,
    };

    const userText1 = "Do dolphins have fingers?";
    console.log("User: ", userText1);
    chatHistory.push({ role: "user", content: userText1 });

    // same as SOC.Union(SOC.ConstString("yes"), SOC.ConstString("no"))
    const yesOrNoGrammar = SOC.Union(SOC.ConstString("yes"), SOC.ConstString("no"));
    generationConfig.structured_output_config = new SOC({ structural_tags_config: yesOrNoGrammar });
    process.stdout.write("Assistant: ");
    const answer1 = await pipe.generate(chatHistory, generationConfig, streamer);
    chatHistory.push({ role: "assistant", content: answer1.texts[0] });

    const userText2 =
        "book flight ticket from Beijing to Paris(using airport code) in 2025-12-04 to 2025-12-10, "
        + "then book hotel from 2025-12-04 to 2025-12-10 in Paris";
    console.log("User: ", userText2);
    chatHistory.push({ role: "user", content: userText2 });

    const startToolCallTag = SOC.ConstString("functools");
    const toolsJson = SOC.JSONSchema(
        toolsToArraySchema(bookFlightTicket, bookHotel)
    );
    const toolCallGrammar = SOC.Concat(startToolCallTag, toolsJson);
    const toolCallParser = new CustomToolCallParser();
    const toolCallStreamer = new ContentStreamer(pipe.getTokenizer(), [toolCallParser]);

    generationConfig.structured_output_config.structural_tags_config = toolCallGrammar;
    generationConfig.parsers = [toolCallParser];

    process.stdout.write("Assistant: ");
    const answer2 = await pipe.generate(chatHistory, generationConfig, (subword) => toolCallStreamer.write(subword));
    console.log();
    console.log(`Finish reason: ${formatFinishReason(answer2.finish_reasons[0])}`);

    console.log("\n\nThe following tool calls were generated:")
    printToolCall(answer2);
    console.log();
}

main();
